comit/src/network/protocols/orderbook.rs

```diff
@@ -357,4 +452,149 @@ impl Orderbook {
+        network::test::{await_events_or_timeout, connect, new_swarm},
+    };
+    use atty::{self, Stream};
+    use libp2p::Swarm;
+    use log::LevelFilter;
+    use spectral::prelude::*;
+    use std::str::FromStr;
+    use tracing::{info, subscriber, Level};
+    use tracing_log::LogTracer;
+    use tracing_subscriber::FmtSubscriber;
+
+    fn create_order(id: PeerId, maker_addr: Multiaddr) -> Order {
+        Order::new(id, NewOrder {
+            buy: asset::Bitcoin::from_sat(100),
+            sell: Erc20 {
+                token_contract: Default::default(),
+                quantity: Erc20Quantity::max_value(),
+            },
+            absolute_expiry: 100,
+            maker_addr,
+        })
+    }
+
+    fn refund_redeem_identities() -> (crate::bitcoin::Address, identity::Ethereum) {
+        let refund = crate::bitcoin::Address::from_str("2MufZ6LLCqYTvZnCwzfAjSKn26Xcnr19PBE")
+            .expect("failed to parse bitcoin address");
+
+        let redeem = identity::Ethereum::random();
+
+        (refund, redeem)
+    }
+
+    #[tokio::test]
+    async fn take_order_request_confirmation() {
+        let comit_level = LevelFilter::Debug;
+        let upstream_level = LevelFilter::Info;
+        init_tracing(comit_level, upstream_level).expect("failed to init tracing");
+
+        // arrange
+
+        let delay = Duration::from_secs(2);
+
+        let (mut alice_swarm, _, _alice_peer_id) = new_swarm(Orderbook::new);
+        let (mut bob_swarm, bob_addr, bob_peer_id) = new_swarm(Orderbook::new);
+        connect(&mut alice_swarm, &mut bob_swarm).await;
+
+        let order = create_order(bob_peer_id.clone(), bob_addr.clone());
+        let (bob_refund_identity, bob_redeem_identity) = refund_redeem_identities();
+
+        // act
+
+        // Trigger subscription.
+        poll_with_delay(&mut alice_swarm, delay).await;
+        poll_with_delay(&mut bob_swarm, delay).await;
+
+        let _ = bob_swarm
+            .make(order.clone(), bob_refund_identity, bob_redeem_identity)
+            .expect("order id should exist");
+
+        // Trigger publish and receipt of order.
+        poll_with_delay(&mut bob_swarm, delay).await;
+        poll_with_delay(&mut alice_swarm, delay).await;
+
+        let order = alice_swarm
+            .get_orders()
+            .first()
+            .cloned()
+            .expect("Alice has no orders");
+
+        tracing::info!("alice received order");
+
+        alice_swarm.take(&bob_peer_id, order.id);
+
+        // Trigger request/response messages.
+        poll_with_delay(&mut bob_swarm, delay).await;
+        poll_with_delay(&mut alice_swarm, delay).await;
+
+        let (alice_event, bob_event) =
+            await_events_or_timeout(alice_swarm.next(), bob_swarm.next()).await;
+
+        // assert
+
+        assert_that(&alice_event).is_equal_to(BehaviourOutEvent::TakeOrderConfirmation);
+        assert_that(&bob_event).is_equal_to(BehaviourOutEvent::TakeOrderRequest);
```

Comment:
:mega: This assertion is a bit weird though. Why does Alice receive a confirmation without Bob explicitly confirming it given that his last emitted event is the TakeOrderRequest?

Reply:
Stale.